### PR TITLE
Product Gallery > Next/Previous Buttons block: Add support to Interactivity API

### LINKED ISSUE #10631: Product Gallery Block: Replace the image in the Large Image Block when utilizing the Next/Previous buttons
## Description
The Nex/Previous block is one of the inner blocks contained within the Product Gallery block. When a shopper navigates on the gallery utilizing the Next/Previous block, the Large Image block should replace its image with the next or previous button that matches

## Outcomes

- The Large Image block should display the image that matches the clicked page from the Nex/Previous block.

## Acceptance criteria
- Upon clicking the Previous from the block, the Large Image block should display the previous image;
- Upon clicking the Next from the block, the Large Image block should display the next image;
- When the image is the last image and upon clicking the Next from the block, no action should occur;
- When the image is the first image and upon clicking the Previous from the block, no action should occur;

### PR DESCRIPTION
**Pull request by thealexandrelara**
## What
This PR adds support to the Interactivity API on the Next/Previous buttons block, allowing shoppers to navigate between the gallery images without having to reload the page.

Fixes #10631 

## Why
This is part of our effort to implement the new Product Gallery block.

## Testing Instructions

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

1. From your WordPress dashboard, go to Appearance > Themes. Make sure you have a block-based theme installed and activated. If not, you can install one from the Add New option. Block-based themes include "Twenty-twenty Two," "Twenty-twenty Three," etc.
2. On the left-hand side menu, click on Appearance > Editor > Templates
3. Find and select the 'Single Product' template from the list.
4. When the Classic Product Template renders, click on Transform into Blocks. This will transform the Classic template in a block template if you haven't done it before.
5. In the block library, search for the 'Product Gallery' block. Click on it to add the block to the template.
6. Check that the Previous/Next button block is displayed on the Editor;
7. Click on save and visit a product page that contains multiple gallery images, make sure the block is correctly being displayed;
8. Make sure you can navigate between the gallery images using the Previous and Next buttons.

* [x] Do not include in the Testing Notes 
* [ ] Should be tested by the development team exclusively 

## Screenshots or screencast

https://github.com/woocommerce/woocommerce-blocks/assets/20469356/afbb167f-88fe-4084-b29b-33fd0f34511b

## WooCommerce Visibility

Required:

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental
* [ ] N/A

### CODE REVIEW

File: assets/js/blocks/product-gallery/frontend.tsx
Comment (gigitux) [inline assets/js/blocks/product-gallery/frontend.tsx]: Usually, the handler function should have the prefix `handle`. What do you think?

```suggestion
			handleNextImageButtonClick: ( store: Store ) => {
```

Also, we should decide if we want a nested structure or not (https://github.com/woocommerce/woocommerce-blocks/pull/10823#discussion_r1325598086). Mmm, I don't have a strong preference, but it makes sense if we have a common approach.

File: assets/js/blocks/product-gallery/frontend.tsx
Comment (gigitux) [inline assets/js/blocks/product-gallery/frontend.tsx]: Same here:
```suggestion
			handlePreviousImageButtonClick: ( store: Store ) => {
```

File: assets/js/blocks/product-gallery/frontend.tsx
Comment (gigitux) [inline assets/js/blocks/product-gallery/frontend.tsx:91]: Good use of `Math.min`!

File: src/BlockTypes/ProductGalleryPager.php
Comment (gigitux) [inline src/BlockTypes/ProductGalleryPager.php]: Is it necessary? It seems that we don't have any frontend files specific to the Product Gallery Pager.

File: src/BlockTypes/ProductGalleryPager.php
Comment (gigitux) [inline src/BlockTypes/ProductGalleryPager.php]: Everything should be inside the woocommerce key. Also, is it necessary `strval` since we converting the ID in the `get_product_gallery_image_ids` function

**Review by gigitux — CHANGES_REQUESTED** [review]
Nice work! The PR works excellent, but I think that there are some things to address. Also, since the Interactivity API is WIP, I think that we should add E2E tests to catch any regressions.

**Comment by gigitux** [comment]
Ops, I just noticed that a lot of code in this PR is in https://github.com/woocommerce/woocommerce-blocks/pull/10736. Please use ZenHub to specify some dependencies between PRs or/and write it in the description PR! 👍

File: assets/js/blocks/product-gallery/frontend.tsx
Comment (thealexandrelara) [reply assets/js/blocks/product-gallery/frontend.tsx]: I just changed it from `onNext/onPrevious` to `handleNext/handlePrevious` so we establish a common pattern to follow

Regarding the nested structure, I believe that since the store is for the entire block (the Product Gallery block), we can leave it flat. However, if there are actions, selectors, or state that should only be used in the context of one of the inner blocks, we can nest them (this way, we make it clear that they should only be used within that specific inner block).

File: src/BlockTypes/ProductGalleryPager.php
Comment (thealexandrelara) [reply src/BlockTypes/ProductGalleryPager.php]: I merged #10736 with this PR so this part was removed because it is unnecessary

**Comment by thealexandrelara** [comment]
> Ops, I just noticed that a lot of code in this PR is in #10736. Please use ZenHub to specify some dependencies between PRs or/and write it in the description PR! 👍

My bad 🤦 , I connected the issue but forgot to add the dependency, thank you for pointing this out and for reviewing the PR!

File: src/BlockTypes/ProductGalleryLargeImageNextPrevious.php
Comment (gigitux) [inline src/BlockTypes/ProductGalleryLargeImageNextPrevious.php:112]: The icons should be different based on the position of the arrow. This change introduces a regression:

| Mockup | Current Implementation |
|--------|--------|
|image|image|
|image|image|

**Review by gigitux — CHANGES_REQUESTED** [review]
Thanks for working on this, and sorry for the late review. During the review, I noticed that a change introduced a regression. I think that we should fix before merging the PR.

File: src/BlockTypes/ProductGalleryLargeImageNextPrevious.php
Comment (thealexandrelara) [reply src/BlockTypes/ProductGalleryLargeImageNextPrevious.php:112]: Nice catch! It was solved in https://github.com/woocommerce/woocommerce-blocks/pull/10938/commits/f6cce8cfe2e0c8617085cd965e71a284d4dfac06

**Review by gigitux — CHANGES_REQUESTED** [review]
Thanks for fixing the arrows on the front end! I noticed that the arrows aren't visible on the editor side:


https://github.com/woocommerce/woocommerce-blocks/assets/4463174/07cca756-ab20-4751-875d-3a5c87de3855

**Comment by thealexandrelara** [comment]
> Thanks for fixing the arrows on the front end! I noticed that the arrows aren't visible on the editor side:

Thank you for pointing this out, I just fixed this in here: https://github.com/woocommerce/woocommerce-blocks/pull/10938/commits/e8a9edb1dfec26645698a9970e17fd133414a5c9

**Review by gigitux — APPROVED** [review]
LGTM!
Thanks for working on this!

**Comment by gigitux** [comment]
I noticed that the PHP linter job step fails. Could you check why, @thealexandrelara?

**Comment by thealexandrelara** [comment]
> I noticed that the PHP linter job step fails. Could you check why, @thealexandrelara?

Yes, for some reason a random empty php file was added to the PR, not sure if it is related to the issue that Github is currently facing with PRs, but after removing this file, it worked

**Pull request merged**
